Put error spans deep on nested object literals

weswigham opened the pull request:
And include an additional location pointing back at the deeply nested property in the target type it is being assigned to. Additionally, like in classes which are incompatible with a base type, object and array literals will now flag _all_ locations which are incompatible with the target instead of just one.

Fixes #25030 partially (for all instances involving expressions where we're tracking which properties have issues).

DanielRosenwasser commented:
As a follow-up to today's error sync, let's use `checkExpressionForMutableLocation`

weswigham commented:
@DanielRosenwasser Done. Also fixed the `enum` and elaboration changes like I said I aught to be able to. Wanna review?

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -10105,8 +10105,137 @@ namespace ts {
+            return false;
+        }
+
+        function *generateObjectliteralElements(node: ObjectLiteralExpression): ElaborationIterator {
```
Capital `L`

DanielRosenwasser commented on src/compiler/diagnosticMessages.json:
```diff
@@ -3710,6 +3710,11 @@
     },
 
+    "Which is from property '{0}' of type '{1}', declared here.": {
+        "category": "Message",
```
I thought it was "category": "Error"

weswigham commented on tests/baselines/reference/deeplyNestedAssignabilityIssue.errors.txt:
```diff
@@ -0,0 +1,67 @@
+    [30;47m  [0m [96m~~~~~~~~~~~~~~~~~~~~~~~~~~~[0m
+    [30;47m13[0m             }
+    [30;47m  [0m [96m~~~~~~~~~~~~~[0m
+    Which is from property '"another"' of type '{ another: A; }', declared here.
```
Because i'm stringifying the literal type associated with the name. It might equally be `'0'` or `[typeof foo]`. I can try to strip `"` from the type's string (if it's a valid identifier post-strip) if you think it'll improve the message.

DanielRosenwasser commented on src/compiler/utilities.ts:
```diff
@@ -6674,6 +6674,25 @@ namespace ts {
+     *
+     * @return non-quoted string
+     */
+    export function stripQuotes(name: string) {
```
I feel like we should just be able to call `symbolToString` and get the right thing in certain cases, and just funnel along a `string`.

weswigham commented:
@DanielRosenwasser I've synced with master, hopefully addressed your reviews, and added a few more tests for good measure. Have anything more you'd like to see?

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -10103,8 +10103,154 @@ namespace ts {
+            if (!length(node.properties)) return;
+            for (const prop of node.properties) {
+                if (isJsxSpreadAttribute(prop)) continue;
+                yield [prop.name, prop.initializer, getLiteralType(idText(prop.name))];
```
It's hard to tell what the intent is here; I'd make this an object instead of a tuple.

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -10103,8 +10103,154 @@ namespace ts {
+            // Assignability failure - check each prop individually, and if that fails, fall back on the bad error span
+            let reportedError = false;
+            for (let status = iterator.next(); !status.done; status = iterator.next()) {
+                const [prop, next, nameType] = status.value;
```
Maybe make it clear that you're diving into `next`; it doesn't seem immediately obvious.

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -10103,8 +10103,154 @@ namespace ts {
+                if (!isTypeAssignableTo(sourcePropType, targetPropType)) {
+                    const elaborated = next && elaborateError(next, sourcePropType, targetPropType);
+                    if (elaborated) {
+                        reportedError = true;
```
I think this can just become a `return`

weswigham replied:
No, if I were to `return` here, i would only issue an elaboration on the first property with an issue instead of all of them.

weswigham replied:
I could `continue` and un-next the `else` if you'd prefer that, however.

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -10103,8 +10103,154 @@ namespace ts {
+        }
+
+        type ElaborationIterator = IterableIterator<[Node, Expression | undefined, Type]>;
+        function elaborateElementwise(iterator: ElaborationIterator, source: Type, target: Type) {
```
Can you write a quick bit about what the intuition should be before I dive into this function?

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -10103,8 +10103,154 @@ namespace ts {
+            return checkTypeRelatedTo(source, target, assignableRelation, errorNode, headMessage, containingMessageChain, errorOutputObject);
+        }
+
+        function checkTypeAssignableToAndOptionallyElaborate(source: Type, target: Type, errorNode: Node | undefined, expr: Expression | undefined, headMessage?: DiagnosticMessage, containingMessageChain?: () => DiagnosticMessageChain | undefined): boolean {
```
Add a little blurb that this performs a structural comparison using a given expression that we can use to provide specific errors on things like properties.

weswigham commented:
>Can you also add a test case similar to what you already had, but which uses a contextual indexed access type?

Ahhh.... to what are you referring?

DanielRosenwasser commented:
Looks like you do have a case - however, as we discussed offline, the computed property message is a little confusing given that it's not clear it you're erroring on the name.

weswigham commented:
@DanielRosenwasser Added; what do you think about the new error?

DanielRosenwasser commented on src/compiler/diagnosticMessages.json:
```diff
@@ -1452,6 +1452,10 @@
         "category": "Error",
         "code": 2417
     },
+    "Type of computed property value is '{0}', which is not assignable to type '{1}'.": {
```
computed property's
